libcontainer/process_linux.go

```diff
@@ -407,6 +407,13 @@ func (p *initProcess) start() (retErr error) {
 		}
 	}()
 
+	// We should join the cgroup after the initial setup finished,
+	// but before runc init clone new children processes. (#4427)
+	err = <-waitInit
+	if err != nil {
+		return err
+	}
+
 	// Do this before syncing with child so that no children can escape the
 	// cgroup. We don't need to worry about not doing this and not being root
 	// because we'd be using the rootless cgroup manager in that case.
```

Comment:
I think this is creating a race condition, for the reason explained here.

The nsexec.c will start creating the ns and all, while we haven't added the process to the cgroup. I think we need to do add another sync for this.

Reply:
Grr, adding a sync is tricky, because the current initWaiter() assumes only one message on the channel.

Reply:
> I think this is creating a race condition, for the reason explained here.

I don't think so. The init process is waiting bootstrapData:
https://github.com/opencontainers/runc/blob/ea5f8e0445e80dc29acd9b5d57481a0e936af89b/libcontainer/process_linux.go#L428-L430

Reply:
Ohhh, right, there is another read in nsexec that syncs that. Great point, thanks!

The current sync seems a little bit tacit/implicit to my taste. But it is true, it is correct, probably for the setns case too as we will read some data before exec into the user process.